src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -91,34 +106,153 @@ def self.compile_ast(code, compile_options, boltlib_path)
+                                    facts: compile_options['facts']['values'],
+                                    variables: compile_options['variables']['values']) do |pal|
+          Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+
+          # This compiler has been configured with a node containing
+          # the requested environment, facts, and variables, and is used
+          # to compile a catalog in that context from the supplied AST.
+          pal.with_catalog_compiler do |compiler|
+            # We have to parse the AST inside the compiler block, because it
+            # initializes the necessary type loaders for us.
+            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
+
+            compiler.evaluate(ast)
+            compiler.compile_additions
+            compiler.catalog_data_hash
+          end
+        end
+      end
+      private_class_method :compile_for_environment
+
+      # Prior to PE-29443 variables were in a hash. Serialization between ruby/clojure/json did
+      # not preserve hash order necessary for deserialization. The data strucutre is now stored
+      # in a list for moving data and the list is used to construct an ordered ruby hash.
+      def self.ordered_plan_vars(compile_options)
+        if compile_options['variables']['values'].is_a?(Array)
+          compile_options['variables']['values'].each_with_object({}) do |param_hash, acc|
+            acc[param_hash.keys.first] = param_hash.values.first
+          end
+        else
+          compile_options['variables']['values']
+        end
+      end
+      private_class_method :ordered_plan_vars
+
+      # Create a puppet settings context that can compile for projects
+      def self.with_puppet_settings_for_projects(overrides = {}, multithreaded)
+        unless multithreaded
+          # Create a copy of the existing settings object
+          original_settings = Puppet.settings
+        end
+
+        Dir.mktmpdir('bolt_compile') do |dir|
+          # This config sets all required settings to an empty dir
+          # where puppet won't load things like puppet.conf
+          puppet_config = {}
+          Puppet::Settings::REQUIRED_APP_SETTINGS.each do |setting|
+            puppet_config[setting] = dir
+          end
+
+          # Actually create the settings context
+          if multithreaded
+            # run generate_standalone_local_settings to pass in the empty
+            # puppet config to the thread local settings object
+            Puppet.settings.generate_standalone_local_settings(puppet_config)
+          else
+            # Replace the existing puppet settings object with one that has 'empty' configuration
+            # (i.e. configuration pointed at an empty dir where puppet won't load things like
+            # puppet.conf)
+            Puppet.replace_settings_object(Puppet::Server::Settings.new(global_settings: [],
```

Comment:
I'm not sure that using this is a viable strategy here, but I don't know with 100% certainty.

When it is used in the `Puppet::Server::Master` class, it is only used in initialization, before we serve any catalogs. In that case, we know it is safe to change out the settings object because it is not accessed by anything else at that point.

However, that is not the case for usage with the compile endpoint; concurrent catalog requests could be coming in. I believe that the changing the settings object will affect other jruby-workers from the pool, leaking the config used for a particular project to other requests.

fwiw, here is the note I left in puppet repo:
https://github.com/puppetlabs/puppet/blob/2cca41ff4911b8822227f64a341be5c6ac327191/lib/puppet.rb#L67-L73

Reply:
My understanding was that in single threaded mode each jruby instance's environment (including the settings for puppet) are completely isolated from other jruby instances. That's why I figured I could use this here: since in single threaded mode the AST compile should be using a jruby instance isolated from other jruby instances we can temporarily replace the settings object because nothing else should be able to use the instance while the compile has control of it's jruby.

Reply:
For multithreaded it uses different logic that leverages `generate_standalone_local_settings` to put in the settings we want in the same way catalog requests do in multithreaded mode.

Reply:
Yeah, I think you are right; as long as the jruby instance gets put back the same way it was received, I think we will be good. However, as written now, exceptions bubbling up through the `yield` will block the way the settings are put back at the end of this method; I think it should be wrapped in an ensure block?

Reply:
Yeah, there's an ensure there already: https://github.com/puppetlabs/puppetserver/pull/2555/files#diff-6f7f4ebbd198c182ba9b89fe43839b117bc1899b5bb3f7ffeddcd1cc64cd79abR244